### PR TITLE
Add load state in <StateContextProvider/> and disable UI elements based on the state.

### PR DESCRIPTION
**Pull request by junhaoliao**
# References

new-log-viewer series: #45 #46 #48 #51 #52 #53 #54 #55 #56 #59 #60 #61 #62 #63 #66 #67 #68 #69 #70 #71 #72 #73 #74 #76 #77 #78 #79 #80 #81 #82 #83 #84

# Description

1. Add load state in <StateContextProvider/>, which includes:
   - UNOPENED 
   - LOADING
   - READY
2. Disable UI elements because of the state in the context.
 
# Validation performed

1. Referred to #46, launched debug server without any files loaded.
2. Performed below actions and observed UI elements disabled / stayed responsive / changed text depending on the state:

|                                                                  | Progress bar | Page-navigation buttons | Page-navigation input box | Open-file button | Log-level Select Button | Log-level Options + Clear Buttons |         Export-logs button         | LogEventNum Button |            Drag-n-drop            |
|:----------------------------------------------------------------:|:------------:|:-----------------------:|:-------------------------:|:----------------:|-------------------------|:---------------------------------:|:----------------------------------:|:------------------:|:---------------------------------:|
| Opened app without any file loaded                               | HIDDEN       | DISABLED                | DISABLED                  | RESPONSIVE       | RESPONSIVE              | DISABLED                          | DISABLED                           | DISABLED           | "Drop file to view"               |
| Opened app with filePath=/valid/file - loading                   | VISIBLE      | DISABLED                | DISABLED                  | DISABLED         | RESPONSIVE              | DISABLED                          | DISABLED                           | RESPONSIVE         | "Drop is disabled during loading" |
| Opened app with filePath=/valid/file - loaded                    | HIDDEN       | RESPONSIVE              | RESPONSIVE                | RESPONSIVE       | RESPONSIVE              | RESPONSIVE                        | RESPONSIVE                         | RESPONSIVE         | "Drop file to view"               |
| Switched page - loading                                          | VISIBLE      | DISABLED                | DISABLED                  | DISABLED         | RESPONSIVE              | DISABLED                          | DISABLED                           | RESPONSIVE         | "Drop is disabled during loading" |
| Switched page - loaded                                           | HIDDEN       | RESPONSIVE              | RESPONSIVE                | RESPONSIVE       | RESPONSIVE              | RESPONSIVE                        | RESPONSIVE                         | RESPONSIVE         | "Drop file to view"               |
| Changed log level - loading                                      | VISIBLE      | DISABLED                | DISABLED                  | DISABLED         | RESPONSIVE              | DISABLED                          | DISABLED                           | RESPONSIVE         | "Drop is disabled during loading" |
| Changed log level - loaded                                       | HIDDEN       | RESPONSIVE              | RESPONSIVE                | RESPONSIVE       | RESPONSIVE              | RESPONSIVE                        | RESPONSIVE                         | RESPONSIVE         | "Drop file to view"               |
| Clicked Export-logs button - loading                             | HIDDEN       | RESPONSIVE              | RESPONSIVE                | RESPONSIVE       | RESPONSIVE              | RESPONSIVE                        | DISABLED; progress display visible | RESPONSIVE         | "Drop file to view"               |
| Clicked Export-logs button - loading - -> dropped file - loading | VISIBLE      | DISABLED                | DISABLED                  | DISABLED         | RESPONSIVE              | DISABLED                          | DISABLED                           | RESPONSIVE         | "Drop is disabled during loading" |
| Clicked Export-logs button - loading - -> dropped file - loaded  | HIDDEN       | RESPONSIVE              | RESPONSIVE                | RESPONSIVE       | RESPONSIVE              | RESPONSIVE                        | RESPONSIVE                         | RESPONSIVE         | "Drop file to view"               |

## Release Notes

- **New Features**
	- Enhanced drag-and-drop functionality with improved user feedback during file operations.
	- Conditional disabling of buttons and inputs based on loading state, providing better user experience.
	- New CSS classes for visual feedback on loading and disabled states across various components.
	- Introduction of `UI_STATE` and `UI_ELEMENT` enumerations for better state management.

- **Bug Fixes**
	- Adjusted styles to resolve layout issues and improve the appearance of UI elements.

- **Documentation**
	- Updated context provider to manage and communicate UI states effectively throughout the application.

### CODE REVIEW

File: new-log-viewer/src/typings/worker.ts [hunk 1]
```diff
@@ -9,6 +9,32 @@ import {
...
+
+    /**
+     * When the file is loaded from any request.
+     */
+    READY = "ready",
+
+    /**
+     * When the file is being exported.
+     */
+    EXPORTING = "exporting",
```
Comment (junhaoliao) [inline new-log-viewer/src/typings/worker.ts:35]: As discussed offline, this log export state can be too specific. It might be a better idea to generalize this.

**Review by davemarco — COMMENTED** [review]
I am not reviewing as i added too much of own code

File: new-log-viewer/src/utils/states.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+/**
+ * Whether a UI element is disabled based on provided state.
+ *
+ * @param uiState
+ * @param uiElement
+ * @return `true` if the element is disabled, `false` otherwise.
+ */
+const isDisabled = (uiState: UI_STATE, uiElement: UI_ELEMENT): boolean => {
+    return false === uiStateGrid[uiState][uiElement];
+};
```
Comment (junhaoliao) [inline new-log-viewer/src/utils/states.ts:116]: 1. ```suggestion
   const isDisabled = (uiState: UI_STATE, uiElement: UI_ELEMENT): boolean => (
       false === uiStateGrid[uiState][uiElement]
   );
   ```


2. The `boolean` return type can also be omitted since it can be inferred but i'll let you decide whether to keep it.

File: new-log-viewer/src/utils/states.ts [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+ * Returns a css class that ignores pointer events if in fast loading state.
+ *
+ * @param uiState
+ * @return The class name or an empty string.
+ */
+const ignoreClicksIfFastLoading = (uiState: UI_STATE): string => {
+    return (uiState === UI_STATE.FAST_LOADING ?
+        "disable-pointer-events" :
+        "");
+};
```
Comment (junhaoliao) [inline new-log-viewer/src/utils/states.ts:128]:
```suggestion
const ignoreClicksIfFastLoading = (uiState: UI_STATE): string => (
    uiState === UI_STATE.FAST_LOADING ?
        "disable-pointer-events" :
        ""
);
```
1. Shall we name it `ignorePointerIfFastLoading` since hovering is also ignored?
2. Similarly the return type can be inferred but feel free to keep it.

File: new-log-viewer/src/index.css [hunk 4]
```diff
@@ -30,6 +30,11 @@ html {
...
     --ylv-menu-bar-z-index: 6;
     --ylv-status-bar-z-index: 6;
+    --ylv-loading-progress-z-index: 9;
     --ylv-drop-file-container-hover-mask-z-index: 10;
     --ylv-drop-file-container-hover-message-z-index: 11;
 }
+
+.disable-pointer-events {
+    pointer-events: none !important;
+}
```
Comment (junhaoliao) [inline new-log-viewer/src/index.css:40]:
```suggestion
}

```

File: new-log-viewer/src/components/MenuBar/PageNumInput.tsx [hunk 5]
```diff
@@ -78,14 +85,18 @@ const PageNumInput = () => {
...
+                disabled={disabled}
                 size={"sm"}
                 slotProps={{input: {ref: inputRef}}}
                 type={"number"}
                 endDecorator={
                     <Typography
-                        className={"page-num-input-num-pages-text"}
                         level={"body-md"}
+
+                        className={`page-num-input-num-pages-text ${disabled ?
```
Comment (junhaoliao) [inline new-log-viewer/src/components/MenuBar/PageNumInput.tsx:97]:
```suggestion
                        className={`page-num-input-num-pages-text ${disabled ?
```

**Review by junhaoliao — COMMENTED** [review]
lgtm with some nits. can't approve from my end since i'm the PR author. @Henry8192 please help take a look.

**Review by davemarco — DISMISSED** [review]

File: new-log-viewer/src/utils/states.ts [hunk 6]
```diff
@@ -0,0 +1,135 @@
+/**
+ * Various states of the UI. UI components may be enabled or disabled based on the state.
+ */
+enum UI_STATE {
```
Comment (Henry8192) [inline new-log-viewer/src/utils/states.ts:4]: I would stay with LOAD_STATE, since these states are all about file loading status.

**Comment by Henry8192** [comment]
A few suggestions:
1. OPEN_FILE_BUTTON and DRAG_AND_DROP should be combined to OPEN_FILE, as they are just two ways of opening files.
2. We can reduce states down to 3: UNOPENED, LOADING & READY. On LOADING state, which includes file opening and page switching, and logs exporting, every component except the progress bar should be disabled. We might introduce another log exporting state, but it is currently out of scope, so let's keep the state simple.

TL;DR:
1. Combine DRAG_AND_DROP and OPEN_FILE_BUTTON to one state.
2. Remove FAST_LOADING and SLOW_LOADING states.

File: new-log-viewer/src/typings/states.ts [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        [UI_ELEMENT.DRAG_AND_DROP]: false,
+    },
+    [UI_STATE.FAST_LOADING]: {
+        [UI_ELEMENT.PROGRESS_BAR]: true,
+        [UI_ELEMENT.NAVIGATION_BAR]: true,
+        [UI_ELEMENT.OPEN_FILE_BUTTON]: true,
+        [UI_ELEMENT.LOG_LEVEL_FILTER]: true,
+        [UI_ELEMENT.EXPORT_LOGS_BUTTON]: true,
+        [UI_ELEMENT.LOG_EVENT_NUM_DISPLAY]: true,
+        [UI_ELEMENT.DRAG_AND_DROP]: false,
```
Comment (junhaoliao) [inline new-log-viewer/src/typings/states.ts:86]: Let's keep the behaviour consistent with `OPEN_FILE_BUTTON`
```suggestion
        [UI_ELEMENT.DRAG_AND_DROP]: true,
```

File: new-log-viewer/src/utils/states.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+ * @param uiState
+ * @param uiElement
+ * @return `true` if the element is disabled, `false` otherwise.
+ */
+const isDisabled = (uiState: UI_STATE, uiElement: UI_ELEMENT): boolean => (
+    false === UI_STATE_GRID[uiState][uiElement]
+);
+
+/**
+ * Returns a css class that ignores pointer events if in fast loading state.
```
Comment (junhaoliao) [inline new-log-viewer/src/utils/states.ts:20]:
```suggestion
 * Returns a CSS class that ignores pointer events if in fast loading state.
```

File: new-log-viewer/src/typings/states.ts [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+ * Stateful elements in the UI.
+ */
+enum UI_ELEMENT {
+    PROGRESS_BAR,
+    NAVIGATION_BAR,
+    OPEN_FILE_BUTTON,
+    LOG_LEVEL_FILTER,
+    EXPORT_LOGS_BUTTON,
+    LOG_EVENT_NUM_DISPLAY,
+    DRAG_AND_DROP,
```
Comment (junhaoliao) [inline new-log-viewer/src/typings/states.ts:44]: if alphabetizing those makes sense
```suggestion
    DRAG_AND_DROP,
    EXPORT_LOGS_BUTTON,
    LOG_EVENT_NUM_DISPLAY,
    LOG_LEVEL_FILTER,
    NAVIGATION_BAR,
    OPEN_FILE_BUTTON,
    PROGRESS_BAR,
```

File: new-log-viewer/src/typings/states.ts [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+ */
+const UI_STATE_GRID: UiStateGridType = Object.freeze({
+    [UI_STATE.UNOPENED]: {
+        [UI_ELEMENT.PROGRESS_BAR]: false,
+        [UI_ELEMENT.NAVIGATION_BAR]: false,
+        [UI_ELEMENT.OPEN_FILE_BUTTON]: true,
+        [UI_ELEMENT.LOG_LEVEL_FILTER]: false,
+        [UI_ELEMENT.EXPORT_LOGS_BUTTON]: false,
+        [UI_ELEMENT.LOG_EVENT_NUM_DISPLAY]: false,
+        [UI_ELEMENT.DRAG_AND_DROP]: true,
```
Comment (junhaoliao) [inline new-log-viewer/src/typings/states.ts:68]: if alphabetizing those makes sense
```suggestion
        [UI_ELEMENT.DRAG_AND_DROP]: true,
        [UI_ELEMENT.EXPORT_LOGS_BUTTON]: false,
        [UI_ELEMENT.LOG_EVENT_NUM_DISPLAY]: false,
        [UI_ELEMENT.LOG_LEVEL_FILTER]: false,
        [UI_ELEMENT.NAVIGATION_BAR]: false,
        [UI_ELEMENT.OPEN_FILE_BUTTON]: true,
        [UI_ELEMENT.PROGRESS_BAR]: false,
```

File: new-log-viewer/src/typings/states.ts [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+    },
+    [UI_STATE.FILE_LOADING]: {
+        [UI_ELEMENT.PROGRESS_BAR]: true,
+        [UI_ELEMENT.NAVIGATION_BAR]: false,
+        [UI_ELEMENT.OPEN_FILE_BUTTON]: false,
+        [UI_ELEMENT.LOG_LEVEL_FILTER]: false,
+        [UI_ELEMENT.EXPORT_LOGS_BUTTON]: false,
+        [UI_ELEMENT.LOG_EVENT_NUM_DISPLAY]: false,
+        [UI_ELEMENT.DRAG_AND_DROP]: false,
```
Comment (junhaoliao) [inline new-log-viewer/src/typings/states.ts:77]: if alphabetizing those makes sense
```suggestion
        [UI_ELEMENT.DRAG_AND_DROP]: false,
        [UI_ELEMENT.EXPORT_LOGS_BUTTON]: false,
        [UI_ELEMENT.LOG_EVENT_NUM_DISPLAY]: false,
        [UI_ELEMENT.LOG_LEVEL_FILTER]: false,
        [UI_ELEMENT.NAVIGATION_BAR]: false,
        [UI_ELEMENT.OPEN_FILE_BUTTON]: false,
        [UI_ELEMENT.PROGRESS_BAR]: true,
```

File: new-log-viewer/src/typings/states.ts (hunk 7 above)
Comment (junhaoliao) [inline new-log-viewer/src/typings/states.ts:86]: if alphabetizing those makes sense
+ note this also includes above `[UI_ELEMENT.DRAG_AND_DROP]: false` -> `true` change
```suggestion
        [UI_ELEMENT.DRAG_AND_DROP]: true,
        [UI_ELEMENT.EXPORT_LOGS_BUTTON]: true,
        [UI_ELEMENT.LOG_EVENT_NUM_DISPLAY]: true,
        [UI_ELEMENT.LOG_LEVEL_FILTER]: true,
        [UI_ELEMENT.NAVIGATION_BAR]: true,
        [UI_ELEMENT.OPEN_FILE_BUTTON]: true,
        [UI_ELEMENT.PROGRESS_BAR]: true,
```

File: new-log-viewer/src/typings/states.ts [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+    },
+    [UI_STATE.SLOW_LOADING]: {
+        [UI_ELEMENT.PROGRESS_BAR]: false,
+        [UI_ELEMENT.NAVIGATION_BAR]: false,
+        [UI_ELEMENT.OPEN_FILE_BUTTON]: false,
+        [UI_ELEMENT.LOG_LEVEL_FILTER]: false,
+        [UI_ELEMENT.EXPORT_LOGS_BUTTON]: false,
+        [UI_ELEMENT.LOG_EVENT_NUM_DISPLAY]: true,
+        [UI_ELEMENT.DRAG_AND_DROP]: false,
```
Comment (junhaoliao) [inline new-log-viewer/src/typings/states.ts:95]:
```suggestion
        [UI_ELEMENT.DRAG_AND_DROP]: false,
        [UI_ELEMENT.EXPORT_LOGS_BUTTON]: false,
        [UI_ELEMENT.LOG_EVENT_NUM_DISPLAY]: true,
        [UI_ELEMENT.LOG_LEVEL_FILTER]: false,
        [UI_ELEMENT.NAVIGATION_BAR]: false,
        [UI_ELEMENT.OPEN_FILE_BUTTON]: false,
        [UI_ELEMENT.PROGRESS_BAR]: false,
```

File: new-log-viewer/src/typings/states.ts [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+    },
+    [UI_STATE.READY]: {
+        [UI_ELEMENT.PROGRESS_BAR]: false,
+        [UI_ELEMENT.NAVIGATION_BAR]: true,
+        [UI_ELEMENT.OPEN_FILE_BUTTON]: true,
+        [UI_ELEMENT.LOG_LEVEL_FILTER]: true,
+        [UI_ELEMENT.EXPORT_LOGS_BUTTON]: true,
+        [UI_ELEMENT.LOG_EVENT_NUM_DISPLAY]: true,
+        [UI_ELEMENT.DRAG_AND_DROP]: true,
```
Comment (junhaoliao) [inline new-log-viewer/src/typings/states.ts:104]:
```suggestion
        [UI_ELEMENT.DRAG_AND_DROP]: true,
        [UI_ELEMENT.EXPORT_LOGS_BUTTON]: true,
        [UI_ELEMENT.LOG_EVENT_NUM_DISPLAY]: true,
        [UI_ELEMENT.LOG_LEVEL_FILTER]: true,
        [UI_ELEMENT.NAVIGATION_BAR]: true,
        [UI_ELEMENT.OPEN_FILE_BUTTON]: true,
        [UI_ELEMENT.PROGRESS_BAR]: false,
```

**Review by junhaoliao — COMMENTED** [review]
some code style suggestions otherwise everything else lgtm

File: new-log-viewer/src/typings/states.ts (hunk 7 above)
Comment (junhaoliao) [reply new-log-viewer/src/typings/states.ts]: I think whenever we drop, we call the `loadFile` method which kills the existing worker, right? Do we still have the multiple inflight requests issue in that case?

Comment (junhaoliao) [reply new-log-viewer/src/typings/states.ts]: > we switch into load file state, then page load comes back before we kill worker

Considering JS's single-threaded nature, that should not happen. i.e., Once we're in `loadFile` nothing else outside would be executed till `loadFile` returns; when `loadFile` returns, we have killed the old worker in `loadFile` so `handleMainWorkerResp` won't run for the old worker even if any messages have been delivered and are pending execution.

File: new-log-viewer/src/contexts/StateContextProvider.tsx [hunk 14]
```diff
@@ -417,6 +434,15 @@ const StateContextProvider = ({children}: StateContextProviderProps) => {
         numPagesRef.current = numPages;
     }, [numPages]);
 
+    // Synchronize `uiStateRef` with `uiState`.
+    useEffect(() => {
+        uiStateRef.current = uiState;
+        if (uiState === UI_STATE.UNOPENED) {
+            setFileName(STATE_DEFAULT.fileName);
+            setLogData(STATE_DEFAULT.logData);
```
[the remark this answers is not in the document]
Comment (junhaoliao) [reply new-log-viewer/src/contexts/StateContextProvider.tsx:442]: Sorry for the confusion, I meant to say
> I suppose we're not resetting the other states **variables** in StateContextProvider

yeah you have answered it. Even if there's a previous successful load, all other state variables like numEvents should have been reset in `loadFile` before we start loading with a new worker.

File: new-log-viewer/src/typings/states.ts [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+        [UI_ELEMENT.DRAG_AND_DROP]: false,
+        [UI_ELEMENT.EXPORT_LOGS_BUTTON]: false,
+        [UI_ELEMENT.LOG_EVENT_NUM_DISPLAY]: false,
+        [UI_ELEMENT.LOG_LEVEL_FILTER]: false,
+        [UI_ELEMENT.NAVIGATION_BAR]: false,
+        [UI_ELEMENT.OPEN_FILE_BUTTON]: false,
+        [UI_ELEMENT.PROGRESS_BAR]: true,
+    },
+    [UI_STATE.FAST_LOADING]: {
+        [UI_ELEMENT.DRAG_AND_DROP]: false,
```
Comment (junhaoliao) [inline new-log-viewer/src/typings/states.ts:80]:
```suggestion
        [UI_ELEMENT.DRAG_AND_DROP]: true,
```

**Review by davemarco — APPROVED** [review]
I approve...
